block-layer/codecs/dag-jose.md

```diff
@@ -0,0 +1,126 @@
+# Specification: DAG-JOSE
+
+**Status: Descriptive - Draft**
+
+JOSE is a stanard for signing and encrypting JSON objects. The various specifications for JOSE can be found in the [IETF datatracker](https://datatracker.ietf.org/wg/jose/documents/). 
+
+DAG JOSE supports the full [IPLD Data Model](../data-model-layer/data-model.md) (within the payload).
+
+## Format
+
+The are two main ways to represent a JOSE object. As a JWS ([json web signature](https://datatracker.ietf.org/doc/rfc7515/?include_text=1)) and JWE ([json web encryption](https://datatracker.ietf.org/doc/rfc7516/?include_text=1)). These two formats  acts as the primitives in JOSE and can be used to create JWT and JWM objects etc. This specification describes how to encode JWS and JWE as an IPLD format.
+
+### Representation
+
+The layout of a decoded JOSE object is described by the IPLD schema defined below. We will refer to this layout as the `Decoded Representation`. 
+
+```ipldsch
+type Signature struct {
+  header optional {String:Any}
+  protected optional {String:Any}
+  signature Bytes
+}
+
+type JWS struct {
+  payload Any
+  signatures [Signature]
+}
+
+type Recipient struct {
+  encrypted_key optional Bytes
+  header optional {String:Any}
+}
+
+type JWE struct {
+  aad optional Bytes
+  ciphertext Bytes
+  iv optional Bytes
+  protected optional {String:Any}
+  recipients [Recipient]
+  tag optional Bytes
+  unprotected optional {String:Any}
+}
+
+type JOSE union {
+  | JWS jws
+  | JWE jwe
+} representation kinded
```

Comment:
This bit with the union doesn't quite "compile".  When using a [kinded representation](https://specs.ipld.io/schemas/representations.html#union-kinded-representation), the only keywords permissible in the union block are kind keywords (int, map, etc), and those must also match the kind of the types they're tagging.

The `JWS` and `JWE` types are both structs, and having no representation stated, structs default to being a map representation.  So kinded unions can't be applied here anyway.

You probably want one of `keyed`, `envelope`, or `inline` representations for this union.

I always advocate for `keyed` wherever possible, because tend to be the friendliest to parser performance; `envelope` as second choice because it's at least still very clear; and `inline` is my least favorite because it's just the one with the most edge cases around it by far (it's easy to see if they'll affect any given situation or not, but personally, I just prefer to make a habit of not using them).

But the choice of which of these seems suitable is up to you.

Reply:
... I've written the above naively assuming that this part of the format is something you have any control over.

At any rate, I do hope we can find some clear and explicit mechanism which disambiguates which of these two types of data is in a block.

Reply:
In discussion in another band, @oed linked me to https://www.rfc-editor.org/rfc/rfc7515.html#section-7 and https://www.rfc-editor.org/rfc/rfc7516.html#section-7 , and indicates that we do indeed have limited control over the format here if we're complying with those existing specs.

I'm writing up a new issue to describe what it seems like we might be wanting from union representations if we were to describe this stuff in IPLD Schemas: https://github.com/ipld/specs/issues/278

However, it's also possible that we just... won't describe this top level union in IPLD Schemas, and will just have to talk about this detail (... at length) and how to disambiguate these structures in prose in this doc instead.  That would be a bit unfortunate, because it's just, well, _more prose_... but that's okay.  We can do that.  We're trying to use schemas here for clarity of docs (and I'm glad we did, because it's highlighted this issue nicely), but there's no technical mandate that we have to have a schema if this is going to be implemented as a codec (nor would there be if it was implemented as an ADL, either).

Reply:
Yeah I think in this case, for now, it's a descriptive device and should just be accompanied by docs to clarify it (either as something inline or with an inline comment pointing to something else). I had to do similar things in the Bitcoin docs I'm working on because I can't control the format but want to use Schemas as a documentation device. I have unions that we can't deal with yet so I called them "faux-unions" in the docs.